rskj-core/src/main/java/co/rsk/peg/RepositoryBtcBlockStoreWithCache.java

```diff
@@ -0,0 +1,214 @@
+        StoredBlock chainHead =  getChainHead();
+        int depth = chainHead.getHeight() - height;
+
+        return getStoredBlockAtMainChainDepth(depth);
+    }
+
+    @Override
+    public StoredBlock getStoredBlockAtMainChainDepth(int depth) throws BlockStoreException {
+        StoredBlock chainHead =  getChainHead();
+        Sha256Hash blockHash = chainHead.getHeader().getHash();
+
+        for (int i = 0; i < depth && blockHash != null; i++) {
+            //If its older than cache go to disk
+            StoredBlock currentBlock = getFromCache(blockHash);
+            if(currentBlock == null) {
+                currentBlock = get(blockHash);
+                if (currentBlock == null) {
+                    return null;
+                }
+            }
+            blockHash = currentBlock.getHeader().getPrevBlockHash();
+        }
+
+        if (blockHash == null) {
+            return null;
+        }
+        StoredBlock block = getFromCache(blockHash);
+        if(block == null) {
+            block = get(blockHash);
+        }
+        int expectedHeight = chainHead.getHeight() - depth;
+        if (block != null && block.getHeight() != expectedHeight) {
+            throw new BlockStoreException(String.format("Block %s at depth %d Height is %d but should be %d",
+                    block.getHeader().getHash(),
+                    depth,
+                    block.getHeight(),
+                    expectedHeight));
+        }
+
+
+        return block;
+    }
+
+    private synchronized void populateCache(StoredBlock chainHead) {
+        cacheBlocks.put(chainHead.getHeader().getHash(), chainHead);
+        Sha256Hash blockHash = chainHead.getHeader().getPrevBlockHash();
+        int depth = MAX_DEPTH_STORED_BLOCKS-1;
+        while (blockHash != null && depth > 0) {
+            if(cacheBlocks.get(blockHash) != null) {
+                break;
+            }
+            StoredBlock currentBlock = get(blockHash);
+            if (currentBlock == null) {
+                break;
+            }
+            cacheBlocks.put(currentBlock.getHeader().getHash(), currentBlock);
+            depth--;
+            blockHash = currentBlock.getHeader().getPrevBlockHash();
+        }
+    }
+
+    private byte[] storedBlockToByteArray(StoredBlock block) {
+        ByteBuffer byteBuffer = ByteBuffer.allocate(128);
+        block.serializeCompact(byteBuffer);
+        byte[] ba = new byte[byteBuffer.position()];
+        byteBuffer.flip();
+        byteBuffer.get(ba);
+        return ba;
+    }
+
+    private StoredBlock byteArrayToStoredBlock(byte[] ba) {
+        ByteBuffer byteBuffer = ByteBuffer.wrap(ba);
+        return StoredBlock.deserializeCompact(btcNetworkParams, byteBuffer);
+    }
+
+    public static class Factory implements BtcBlockStoreWithCache.Factory {
+        private final Map<Sha256Hash, StoredBlock> cacheBlocks = new MaxSizeHashMap<>(MAX_SIZE_MAP_STORED_BLOCKS, true);
```

Comment:
The idea is that all block stores created with a specific factory share the same cache. I love it.

One thing that isn't a problem now because we don't have parallel execution is concurrent access to this cache, maybe leaving a comment is worth it.

Reply:
good idea